libraries/AP_Beacon/AP_Beacon_Nooploop.cpp

```diff
@@ -0,0 +1,238 @@
+        }
+        break;
+
+    case ParseState::NF2_PAYLOAD:
+        // add byte to buffer if there is room
+        if (_msg_len < NOOPLOOP_MSG_BUF_MAX) {
+            _msgbuf[_msg_len] = b;
+        }
+        _msg_len++;
+        if (_msg_len >= _frame_len) {
+            _state = ParseState::HEADER;
+            // check crc
+            if (b == _crc_expected) {
+                return MsgType::NODE_FRAME2;
+            } else {
+                return MsgType::INVALID;
+            }
+        } else {
+            _crc_expected += b;
+        }
+        break;
+
+    case ParseState::SF0_PAYLOAD:
+        // add byte to buffer if there is room
+        if (_msg_len < NOOPLOOP_MSG_BUF_MAX) {
+            _msgbuf[_msg_len] = b;
+        }
+        _msg_len++;
+        if (_msg_len >= 128) {
+            _state = ParseState::HEADER;
+            // check crc
+            if (b == _crc_expected) {
+                return MsgType::SETTING_FRAME0;
+            } else {
+                return MsgType::INVALID;
+            }
+        } else {
+            _crc_expected += b;
+        }
+        break;
+    }
+
+    return MsgType::INVALID;
+}
+
+void AP_Beacon_Nooploop::parse_node_frame2()
+{
+    // a message has been received
+    _last_update_ms = AP_HAL::millis();
+
+    // estimated precision for x,y position in meters
+    const float precision_x = _msgbuf[NOOPLOOP_NODE_FRAME2_PRECISION_X] * 0.01;
+    const float precision_y = _msgbuf[NOOPLOOP_NODE_FRAME2_PRECISION_Y] * 0.01;
+    //EKF's estimate goes very bad if the error value sent into the EKF is unrealistically low. ensure it's never less than a reasonable value
+    const float pos_err = MAX(0.1f, sqrtf(sq(precision_x)+sq(precision_y)));
+
+    // x,y,z position in m*1000 in ENU frame
+    const int32_t pos_x = ((int32_t)_msgbuf[NOOPLOOP_NODE_FRAME2_POSX+2] << 24 | (int32_t)_msgbuf[NOOPLOOP_NODE_FRAME2_POSX+1] << 16 | (int32_t)_msgbuf[NOOPLOOP_NODE_FRAME2_POSX] << 8) >> 8;
+    const int32_t pos_y = ((int32_t)_msgbuf[NOOPLOOP_NODE_FRAME2_POSY+2] << 24 | (int32_t)_msgbuf[NOOPLOOP_NODE_FRAME2_POSY+1] << 16 | (int32_t)_msgbuf[NOOPLOOP_NODE_FRAME2_POSY] << 8) >> 8;
+    const int32_t pos_z = ((int32_t)_msgbuf[NOOPLOOP_NODE_FRAME2_POSZ+2] << 24 | (int32_t)_msgbuf[NOOPLOOP_NODE_FRAME2_POSZ+1] << 16 | (int32_t)_msgbuf[NOOPLOOP_NODE_FRAME2_POSZ] << 8) >> 8;
+
+    // position scaled to meters and changed to NED
+    const Vector3f pos_m {pos_y * 0.001f, pos_x * 0.001f, -pos_z * 0.001f};
+
+    set_vehicle_position(pos_m, pos_err);
+
+    const uint8_t valid_nodes = _msgbuf[NOOPLOOP_NODE_FRAME2_VALID_NODES];
+    for (uint8_t i = 0; i < valid_nodes; i++) {
+        uint16_t offset = NOOPLOOP_NODE_FRAME2_NODE_BLOCK + i * 13;
+        uint8_t id = _msgbuf[offset+1]; //nooploop id starts from 0, increments clockwise, 0 -> 1 define Y axis.
+        const int32_t dist = ((int32_t)_msgbuf[offset+2+2] << 24 | (int32_t)_msgbuf[offset+2+1] << 16 | (int32_t)_msgbuf[offset+2] << 8) >> 8;
```

Comment:
Would be nice to have a function which did this.  `uint24_from_buf` or something.